### DISCUSSION TITLE
How to retrieve the absolute coordinates of the predicted bounding boxes

### DISCUSSION THREAD

**Discussion by ghost** [mindee/doctr#570] [Q&A] [answered]
I want to save bounding box coordinates of highlighted text how can I save them
my original image 
![passbook_68_0](https://user-images.githubusercontent.com/70446516/139680893-ecb5cc53-8b7c-4c68-9816-23ea0d4065af.jpeg)
and after prediction

![download (4)](https://user-images.githubusercontent.com/70446516/139680806-93514f68-75be-4bcd-8d94-60b708f5fc85.png)

**Comment by fg-mindee** [comment]
Hi @hammadyounas2008 :wave: 

Would you mind getting a bit more specific? I think you asked the same question in https://github.com/mindee/doctr/discussions/561 and I provided an answer. Please check the discussions & issues to see if there is an answer to your question before posting it :pray: 

If I'm mistaking, please provide more details: python code snippet that you used, what you expect (a saved file locally, a python dict?)

**Reply by ghost** [reply]
Like the blue bounding box shown with blue color on every word, I want to save the coordinates for every word, How can I save that like (x, y, w, h) ? or (xmin, ymin, xmax, ymax)


The following code I am using
```
%cd /content/doctr/doctr
import cv2
from doctr.io import DocumentFile
from doctr.models import ocr_predictor
import matplotlib.pyplot as plt

model = ocr_predictor(pretrained=True)
# PDF
single_img_doc = DocumentFile.from_images("/content/passbook_87_0.jpeg")
# Analyze
result = model(single_img_doc)
print(result)
img=result.show(single_img_doc)

```

**Reply by fg-mindee** [reply]
Could you check my answer in https://github.com/mindee/doctr/discussions/561 please? I think that's precisely what you're asking for, so I'd prefer to avoid duplicating answers and avoid confusion for future visitors :sweat_smile:

**Reply by ghost** [reply]
That's not the answer, did you understand the question I asked?

**Reply by ghost** [reply]
I want to save coordinates values of every word, like Saving, Bank, Account every word has a bounding box on it. I want the coordinates of that bounding box.
img=result.show(single_img_doc)
Only show the predicted image not the bounding box values of every word

**Reply by fg-mindee** [reply]
Well, yes I think I do :sweat_smile: 

And that's why I suggested to check https://github.com/mindee/doctr/discussions/561. As you can see there, I did not mention any `img=result.show(single_img_doc)`, but how to export the results (coordinates & text for each word, line & block) as a Python dict, and how to save it locally as a JSON dictionary.

I thought that would fit your question nicely? If I misunderstood, please give a short example of a typical output you are expecting :pray:

**Reply by ghost** [reply]
You are right I have to save the JSON file, but mainly my question is how can I get bounding box coordinates value in the JSON FIle.
You can say I want to use that coordinate values to extract the word and save that cropped image in a folder.

The JSON file looks like this 

{"value": "Phone", "confidence": 0.9546781778335571, "geometry": [[0.486328125, 0.759765625], [0.541015625, 0.78125]]}, {"value": "or", "confidence": 0.9642261862754822, "geometry": [[0.462890625, 0.7626953125], [0.4873046875, 0.7822265625]]}, {"value": "apdation.", "confidence": 0.16572953760623932, "geometry": [[0.7578125, 0.7587890625], [0.83984375, 0.7802734375]]}, {"value": "Change", "confidence": 0.9972648024559021, "geometry": [[0.25390625, 0.7626953125], [0.3212890625, 0.78515625]]}, {"value": "any", "confidence": 0.9973770380020142, "geometry": [[0.2138671875, 0.7626953125], [0.25390625, 0.7861328125]]}, {"value": ".Guts", "confidence": 0.05273749679327011, "geometry": [[0.689453125, 0.86328125], [0.962890625, 0.986328125]]}]}]

Where is the coordinates values of xmin ,ymin,xmax, ymax, or you can say x,y,w,h

**Reply by fg-mindee** [reply]
Oh that's rather simple then:
- each `geometry` gives you `[[xmin, ymin], [xmax, ymax]]` in relative coordinates format (cf. https://mindee.github.io/doctr/latest/io.html#doctr.io.Word)
- multiply then by the width & height of the page and you'll get the absolute coordinates you are looking for :)

**Reply by ghost** [reply]
Can you please share that script, I am doing it for the past one week but I am unable to do that.

**Reply by fg-mindee** [reply]
Something like this should do:
```python
# Everything to produce the result goes here
# ...
export = result.export()
# Flatten the export
page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in export['pages']]
page_dims = [page['dimensions'] for page in export['pages']]
# Get the coords in [xmin, ymin, xmax, ymax]
words_abs_coords = [
    [[int(round(word['geometry'][0][0] * dims[1])), int(round(word['geometry'][0][1] * dims[0])), int(round(word['geometry'][1][0] * dims[1])), int(round(word['geometry'][1][1] * dims[0]))] for word in words]
    for words, dims in zip(page_words, page_dims)
]
```

Let me know if you're still having troubles!

**Reply by ghost** [reply]
File "<ipython-input-8-ed1fd0e239e1>", line 3
    page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']]] for page in result['pages']]
                                                                                                             ^
SyntaxError: invalid syntax

**Reply by fg-mindee** [reply]
My bad, I fixed the typo, that line should be:
```python
page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in result['pages']]
```

**Reply by ghost** [reply]
<ipython-input-13-ed66b187710e> in <module>()
      1 export = result.export()
      2 # Flatten the export
----> 3 page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in result['pages']]
      4 page_dims = [page['dimensions'] for page in result['pages']]
      5 # Get the coords in [xmin, ymin, xmax, ymax]

TypeError: 'Document' object is not subscriptable

**Reply by fg-mindee** [reply]
My apologies, there were some typos (`results['pages']` --> `export['pages']`) again. I just edited the script on the following lines:
```python
page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in export['pages']]
page_dims = [page['dimensions'] for page in export['pages']]
```
and tried it on my end, it should be working now :+1:

**Reply by ghost** [reply]
@fg-mindee  Thanks Can I draw a bounding box on each line except on words? and save the crop line image in the folder using OpenCV. How can I do that? i have posted that question too.

**Reply by ghost** [reply]
@fg-mindee Can you please reply?

**Reply by fg-mindee** [reply]
Hi @hammadyounas2008 :wave: 

I can see two aspects to address here:

First, let's bear in mind that this is an open source project :hugs:  That has two implications: you can enjoy the source code in whichever you see fit (almost at least, cf. the license), AND this is not a commercial product (there is no customer center, only willing maintainers just like most open source projects). So let's stay mindful of this (cf. [code of conduct](https://github.com/mindee/doctr/blob/main/CODE_OF_CONDUCT.md)), and not expect maintainers to answer within a few hours :grinning: Mindee is opening resources like docTR to the community, but that doesn't mean this is the full-time job of everyone here (or that we don't have other contributors/requests to attend to). _We might not be in the same time zone, but you asked something here at 3:30am Paris time, so it should be understandable that I won't necessarily answer back the next morning :sweat_smile:_ 

Second, in order for us to find the right information later on (or for other visitors), let's try to avoid mixing issues/discussions together :pray: This current discussion was about getting absolute coordinates from the predictions' export, which I think was answered above :) About word aggregation into words, let's discuss this on the other discussion you created!

I hope you don't take it the wrong way, let's try to keep this a friendly & respectful environment for the community :raised_hands: 
Have a good evening, cheers :v:

**Reply by ghost** [reply]
@fg-mindee Sorry for that, I have posted the question but no one response there that's why

**Reply by fg-mindee** [reply]
No worries :ok_hand: 
Did my last edit of the snippet solve your absolute coordinate retrieval problem by the way?

**Reply by ghost** [reply]
No , when I draw the values on the image it does not draw on a specific word

![download (7)](https://user-images.githubusercontent.com/70446516/140168992-a00eb2a2-d784-4a61-a530-5cb02ce6406d.png)
```
export = result.export()
# Flatten the export
page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in export['pages']]
page_dims = [page['dimensions'] for page in export['pages']]
# Get the coords in [xmin, ymin, xmax, ymax]
words_abs_coords = [
    [[int(round(word['geometry'][0][0] * dims[0])), int(round(word['geometry'][0][1] * dims[1])), int(round(word['geometry'][1][0] * dims[0])), int(round(word['geometry'][1][1] * dims[1]))] for word in words]
    for words, dims in zip(page_words, page_dims)

]
print(words_abs_coords)
```
Following are the absolute coordinates 
```
[[[33, 108, 57, 135], [54, 107, 81, 136], [189, 110, 205, 141], [205, 112, 221, 141], [222, 114, 230, 141], [230, 112, 247, 141], [11, 173, 39, 196], [41, 175, 68, 196], [71, 175, 87, 198], [90, 177, 116, 198], [215, 179, 256, 199], [26, 204, 35, 225], [10, 203, 25, 227], [89, 204, 131, 228], [214, 207, 256, 227], [54, 228, 57, 236], [11, 225, 38, 246], [41, 224, 53, 247], [90, 225, 129, 245], [11, 244, 42, 265], [45, 245, 64, 267], [82, 246, 102, 267], [67, 246, 79, 268], [104, 247, 127, 268], [13, 301, 87, 324], [90, 303, 113, 323], [12, 327, 60, 349], [63, 331, 69, 347], [84, 331, 125, 349], [70, 328, 80, 351], [214, 334, 259, 356], [61, 360, 108, 378], [41, 357, 59, 382], [130, 360, 160, 381], [111, 359, 128, 382], [214, 362, 282, 386], [41, 388, 62, 411], [63, 388, 84, 411], [85, 388, 106, 411], [108, 387, 131, 410], [213, 392, 237, 415], [239, 393, 276, 418], [11, 415, 34, 439], [213, 419, 230, 444], [231, 419, 241, 444], [244, 422, 286, 447], [11, 443, 34, 467], [208, 441, 252, 477], [259, 451, 287, 476], [11, 474, 34, 497], [52, 471, 80, 496], [38, 470, 51, 498], [215, 478, 274, 501], [10, 501, 30, 525], [207, 505, 267, 531], [49, 531, 123, 555], [11, 536, 27, 559], [29, 534, 46, 562], [204, 536, 233, 560], [234, 538, 259, 562]]]
```
the following code is used to plot the bounding box but the output I strange it does not draw on the specific letter
```
import matplotlib.pyplot as plt
import cv2
image = cv2.imread("/content/passbook_82_0.jpeg")
im_height, im_width, _ = image.shape
xmin=words_abs_coords[0][0][0]
ymin=words_abs_coords[0][0][1]
xmax=words_abs_coords[0][0][2]
ymax=words_abs_coords[0][0][3]
image1 = cv2.rectangle(image, (xmin,ymin), (xmax,ymax), (0,255,0), 2)
plt.imshow(image1)

```

**Reply by fg-mindee** [reply]
I tried locally it was because of an inversion between page dimensions to generate the abs coords. Now that should work properly :+1:
_(in the conversion to abs coordinates, simply switch `dims[0]` to `dims[1]` and conversely, I edited the snippet)_

**Reply by ghost** [reply]
@fg-mindee  Can i Draw a single bounding box on that line?

**Reply by fg-mindee** [reply]
That's a different topic, and we already have a discussion about this over there :point_right: https://github.com/mindee/doctr/discussions/573 :sweat_smile: 

But the code snippet I provided here in this discussion will not plot a box on lines since we've flattened everything

**Comment by fg-mindee** [comment, marked as answer]
For anyone looking for the answer, here is the snippet from the other thread starting from the predictor result:
```python
# Everything to produce the result goes here
# ...
export = result.export()
# Flatten the export
page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page in export['pages']]
page_dims = [page['dimensions'] for page in export['pages']]
# Get the coords in [xmin, ymin, xmax, ymax]
words_abs_coords = [
    [[int(round(word['geometry'][0][0] * dims[1])), int(round(word['geometry'][0][1] * dims[0])), int(round(word['geometry'][1][0] * dims[1])), int(round(word['geometry'][1][1] * dims[0]))] for word in words]
    for words, dims in zip(page_words, page_dims)
]
```

**Comment by khawar-islam** [comment]
Hello @fg-mindee,

Thank you for your great work. I am simply retrieving a lot of images from folder and apply OCR model to detect all text in an image and save into the another folder. I am getting strange behavior after OCR detection the image is same in another folder and there is not detected part in an image.

**Code**

```
for img in glob.glob(testing_path + '*.*'):
    img_bgr_rgb = cv2.imread(img)
    file_Name = os.path.basename(img)
    image = DocumentFile.from_images(img)
    model = ocr_predictor('db_resnet50', 'sar_resnet31', pretrained=True)
    result = model(image)

    export = result.export()
    # Flatten the export
    page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page
                  in export['pages']]
    page_dims = [page['dimensions'] for page in export['pages']]
    words_abs_coords = [
        [[int(round(word['geometry'][0][0] * dims[1])), int(round(word['geometry'][0][1] * dims[0])),
          int(round(word['geometry'][1][0] * dims[1])), int(round(word['geometry'][1][1] * dims[0]))] for word in
         words]
        for words, dims in zip(page_words, page_dims)
    ]
    xmin = words_abs_coords[0][0][0]
    ymin = words_abs_coords[0][0][1]
    xmax = words_abs_coords[0][0][2]
    ymax = words_abs_coords[0][0][3]

    imagee = cv2.rectangle(img_bgr_rgb, (xmin, ymin), (xmax, ymax), (255, 87, 51), 2)
    cv2.imwrite(os.path.join(result_path, file_Name), img_bgr_rgb)
```

**Reply by frgfm** [reply]
Let's avoid duplication discussion and move the rest of this over there :point_right:  https://github.com/mindee/doctr/discussions/561#discussioncomment-3052591
